crates/oci/src/lib.rs

```diff
@@ -1,448 +1,38 @@
-    pub fn insert(
-        &mut self,
-        server: impl AsRef<str>,
-        username: impl AsRef<str>,
-        password: impl AsRef<str>,
-    ) -> Result<()> {
-        let encoded = base64::Engine::encode(
-            &base64::engine::general_purpose::STANDARD,
-            format!("{}:{}", username.as_ref(), password.as_ref()),
-        );
-        self.auths.insert(server.as_ref().to_string(), encoded);
-
-        Ok(())
-    }
-
-    fn default_path() -> Result<PathBuf> {
-        Ok(dirs::config_dir()
-            .context("Cannot find configuration directory")?
-            .join("fermyon")
-            .join("registry-auth.json"))
-    }
-
-    /// Get the registry authentication for a given registry from the default location.
-    async fn get_auth_from_default(server: impl AsRef<str>) -> Result<RegistryAuth> {
-        let auths = Self::load_default().await?;
-        let encoded = match auths.auths.get(&server.as_ref().to_string()) {
-            Some(e) => e,
-            None => bail!(format!("no credentials stored for {}", server.as_ref())),
-        };
-
-        let bytes = base64::Engine::decode(&base64::engine::general_purpose::STANDARD, encoded)?;
-        let decoded = std::str::from_utf8(&bytes)?;
-        let parts: Vec<&str> = decoded.splitn(2, ':').collect();
-
-        tracing::trace!("Decoded registry credentials from the Spin configuration.");
-        Ok(RegistryAuth::Basic(
-            parts
-                .first()
-                .context("expected username as first element of the decoded auth")?
-                .to_string(),
-            parts
-                .get(1)
-                .context("expected secret as second element of the decoded auth")?
-                .to_string(),
-        ))
-    }
-
-    async fn load(p: &Path) -> Result<Self> {
-        let contents = tokio::fs::read_to_string(&p).await?;
-        serde_json::from_str(&contents)
-            .with_context(|| format!("cannot load authentication file {:?}", p))
-    }
-
-    async fn save(&self, p: &Path) -> Result<()> {
-        tokio::fs::write(&p, &serde_json::to_vec_pretty(&self)?)
-            .await
-            .with_context(|| format!("cannot save authentication file {:?}", p))
-    }
+mod auth;
+mod client;
+
+pub use client::Client;
+
+/// Applies a heuristic to check if the given string "looks like" it could be
+/// an OCI reference.
+pub fn is_probably_oci_reference(maybe_oci: &str) -> bool {
+    // A relative file path such as foo/spin.toml will successfully
+    // parse as an OCI reference, because the parser infers the Docker
+    // registry and the `latest` version.  So if the registry resolves
+    // to Docker, but the source *doesn't* contain the string 'docker',
+    // we can guess this is likely a false positive.
+    //
+    // This could be fooled by, e.g., dockerdemo/spin.toml.  But we only
+    // go down this path if the file does not exist, and the chances of
```

Comment:
Now this is moved out of private `up` code, there's no longer a sense of "this gets called if a string doesn't exist in the file system" - the knowledge that the string is _not_ a file is, I think, important to the heuristic.  So if we move it here (and I agree it's good to concentrate all Spin's understanding of OCI in one place), we maybe need to reword the documentation (and this comment) to clarify that that context...?